### PR TITLE
[FIXED] Check for checksum violations for all records and before any sequence processing.

### PR DESCRIPTION
**Pull request by derekcollison**
Also small bug fix for leaking fds under certain scenarios during corruption.

Signed-off-by: Derek Collison <derek@nats.io>

### CODE REVIEW

**Review by wallyqs — APPROVED** [review]
LGTM

File: server/filestore.go [hunk 1]
```diff
@@ -1078,6 +1078,27 @@ func (mb *msgBlock) rebuildStateLocked() (*LostStreamData, error) {
...
+			hh.Write(hdr[4:20])
+			hh.Write(data[:slen])
+			if hasHeaders {
+				hh.Write(data[slen+4 : dlen-recordHashSize])
+			} else {
+				hh.Write(data[slen : dlen-recordHashSize])
+			}
+			checksum := hh.Sum(nil)
+			if !bytes.Equal(checksum, data[len(data)-recordHashSize:]) {
+				//				fmt.Printf("FAILED CHECKSUM!!!\n\n\n")
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/filestore.go]: Fixed thanks!

**Review by wallyqs — APPROVED** [review]
LGTM

**Review by neilalexander — APPROVED** [review]
LGTM

**Pull request merged**